src/pyqasm/pulse/openpulse_visitor.py

```diff
@@ -0,0 +1,997 @@
+            frame_arg = capture_args[0]
+            if isinstance(frame_arg, pulse_ast.Identifier):
+                _ = self._get_frame(statement, frame_arg.name)
+            else:
+                raise_qasm3_error(
+                    f"Invalid frame argument '{type(frame_arg).__name__}' "
+                    f"in '{statement.name.name}' function",
+                    error_node=statement,
+                    span=statement.span,
+                )
+            waveform_arg = capture_args[1]
+            if isinstance(waveform_arg, pulse_ast.Identifier):
+                if waveform_arg.name not in self._waveforms:
+                    raise_qasm3_error(
+                        f"Waveform '{waveform_arg.name}' not declared",
+                        error_node=statement,
+                        span=statement.span,
+                    )
+            else:
+                raise_qasm3_error(
+                    f"Invalid waveform argument '{type(waveform_arg).__name__}' "
+                    f"in '{statement.name.name}' function",
+                    error_node=statement,
+                    span=statement.span,
+                )
+
+        if statement.name.name in ["capture_v3", "capture_v4"]:
+            capture_args = statement.arguments
+            if len(capture_args) != 2:
+                raise_qasm3_error(
+                    f"Invalid number of arguments for '{statement.name.name}' function",
+                    error_node=statement,
+                    span=statement.span,
+                )
+            self._waveform_validator.validate_capture_v3_and_v4_waveform(statement)
+
+        if self._check_only:
+            return _return_value, []
+
+        statement = qasm3_ast.ExpressionStatement(expression=statement)  # type: ignore
+
+        return _return_value, [statement]
+
+    def _visit_delay_instruction(
+        self, statement: pulse_ast.DelayInstruction
+    ) -> list[pulse_ast.Statement]:
+        """Visit a delay instruction statement.
+
+        Args:
+            statement (pulse_ast.DelayInstruction): The delay instruction statement to visit.
+        Returns:
+            None
+        """
+        self._qasm_visitor.visit_statement(statement)
+        # implicit phase tracking
+
+        return [statement]
+
+    def visit_statement(self, statement: pulse_ast.Statement) -> list[pulse_ast.Statement]:
+        """Visit a statement element.
+
+        Args:
+            statement (pulse_ast.Statement): The statement to visit.
+
+        Returns:
+            None
+        """
+        logger.debug("Visiting statement '%s'", str(statement))
+        result = []
+        visit_map = {
+            pulse_ast.QuantumBarrier: self._visit_barrier,
+            pulse_ast.ClassicalDeclaration: self._visit_classical_declaration,
+            pulse_ast.ExpressionStatement: lambda x: self._visit_function_call(x.expression),
+            pulse_ast.DelayInstruction: self._visit_delay_instruction,
```

Comment:
```suggestion
            pulse_ast.DelayInstruction: self.qasm_visitor._visit_delay_instruction,
```

I think this should work as well